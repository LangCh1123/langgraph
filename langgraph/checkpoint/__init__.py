from langgraph.checkpoint.base import BaseCheckpointSaver, Checkpoint, CheckpointAt
from langgraph.checkpoint.memory import MemorySaver
from langgraph.checkpoint.dynamodb import DynamoDbSaver

__all__ = [
    "Checkpoint",
    "CheckpointAt",
    "BaseCheckpointSaver",
    "MemorySaver",
    "DynamoDbSaver"
]
