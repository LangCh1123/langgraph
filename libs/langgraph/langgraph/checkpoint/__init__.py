from langgraph_checkpoint.base import (
    BaseCheckpointSaver,
    Checkpoint,
    SerializerProtocol,
)
from langgraph_checkpoint.memory import MemorySaver

__all__ = [
    "BaseCheckpointSaver",
    "Checkpoint",
    "MemorySaver",
    "SerializerProtocol",
]
