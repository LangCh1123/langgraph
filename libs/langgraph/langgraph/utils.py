import asyncio
import enum
import inspect
import sys
from contextvars import copy_context
from functools import partial, wraps
from typing import Any, AsyncIterator, Awaitable, Callable, Optional, Type, Union

from langchain_core.runnables.base import (
    Runnable,
    RunnableConfig,
    RunnableLambda,
    RunnableLike,
    RunnableParallel,
)
from langchain_core.runnables.config import (
    merge_configs,
    run_in_executor,
    var_child_runnable_config,
)
from langchain_core.runnables.utils import accepts_config
from typing_extensions import (
    Annotated,
    NotRequired,
    ReadOnly,
    Required,
    TypeGuard,
    get_origin,
)

try:
    from langchain_core.runnables.config import _set_config_context
except ImportError:
    # For forwards compatibility
    def _set_config_context(context: RunnableConfig) -> None:  # type: ignore
        """Set the context for the current thread."""
        var_child_runnable_config.set(context)


# Before Python 3.11 native StrEnum is not available
class StrEnum(str, enum.Enum):
    """A string enum."""


class RunnableCallable(Runnable):
    """A much simpler version of RunnableLambda that requires sync and async functions."""

    def __init__(
        self,
        func: Callable[..., Optional[Runnable]],
        afunc: Optional[Callable[..., Awaitable[Optional[Runnable]]]] = None,
        *,
        name: Optional[str] = None,
        tags: Optional[list[str]] = None,
        trace: bool = True,
        recurse: bool = True,
        **kwargs: Any,
    ) -> None:
        if name is not None:
            self.name = name
        elif func:
            try:
                if func.__name__ != "<lambda>":
                    self.name = func.__name__
            except AttributeError:
                pass
        elif afunc:
            try:
                self.name = afunc.__name__
            except AttributeError:
                pass
        self.func = func
        self.afunc = afunc
        self.config: Optional[RunnableConfig] = {"tags": tags} if tags else None
        self.kwargs = kwargs
        self.trace = trace
        self.recurse = recurse

    def __repr__(self) -> str:
        repr_args = {
            k: v
            for k, v in self.__dict__.items()
            if k not in {"name", "func", "afunc", "config", "kwargs", "trace"}
        }
        return f"{self.get_name()}({', '.join(f'{k}={v!r}' for k, v in repr_args.items())})"

    def invoke(
        self, input: Any, config: Optional[RunnableConfig] = None, **kwargs: Any
    ) -> Any:
        if self.func is None:
            raise TypeError(
                f'No synchronous function provided to "{self.name}".'
                "\nEither initialize with a synchronous function or invoke"
                " via the async API (ainvoke, astream, etc.)"
            )
        kwargs = {**self.kwargs, **kwargs}
        if self.trace:
            ret = self._call_with_config(
                self.func, input, merge_configs(self.config, config), **kwargs
            )
        else:
            config = merge_configs(self.config, config)
            context = copy_context()
            context.run(_set_config_context, config)
            if accepts_config(self.func):
                kwargs["config"] = config
            ret = context.run(self.func, input, **kwargs)
        if isinstance(ret, Runnable) and self.recurse:
            return ret.invoke(input, config)
        return ret

    async def ainvoke(
        self, input: Any, config: Optional[RunnableConfig] = None, **kwargs: Any
    ) -> Any:
        if not self.afunc:
            return self.invoke(input, config)
        kwargs = {**self.kwargs, **kwargs}
        if self.trace:
            ret = await self._acall_with_config(
                self.afunc, input, merge_configs(self.config, config), **kwargs
            )
        else:
            config = merge_configs(self.config, config)
            context = copy_context()
            context.run(_set_config_context, config)
            if accepts_config(self.afunc):
                kwargs["config"] = config
            if sys.version_info >= (3, 11):
                ret = await asyncio.create_task(
                    self.afunc(input, **kwargs), context=context
                )
            else:
                ret = await self.afunc(input, **kwargs)
        if isinstance(ret, Runnable) and self.recurse:
            return await ret.ainvoke(input, config)
        return ret


def is_async_callable(
    func: Any,
) -> TypeGuard[Callable[..., Awaitable]]:
    """Check if a function is async."""
    return (
        asyncio.iscoroutinefunction(func)
        or hasattr(func, "__call__")
        and asyncio.iscoroutinefunction(func.__call__)
    )


def is_async_generator(
    func: Any,
) -> TypeGuard[Callable[..., AsyncIterator]]:
    """Check if a function is an async generator."""
    return (
        inspect.isasyncgenfunction(func)
        or hasattr(func, "__call__")
        and inspect.isasyncgenfunction(func.__call__)
    )


def coerce_to_runnable(thing: RunnableLike, *, name: str, trace: bool) -> Runnable:
    """Coerce a runnable-like object into a Runnable.

    Args:
        thing: A runnable-like object.

    Returns:
        A Runnable.
    """
    if isinstance(thing, Runnable):
        return thing
    elif is_async_generator(thing) or inspect.isgeneratorfunction(thing):
        return RunnableLambda(thing, name=name)
    elif callable(thing):
        if is_async_callable(thing):
            return RunnableCallable(None, thing, name=name, trace=trace)
        else:
            return RunnableCallable(
                thing,
                wraps(thing)(partial(run_in_executor, None, thing)),
                name=name,
                trace=trace,
            )
    elif isinstance(thing, dict):
        return RunnableParallel(thing)
    else:
        raise TypeError(
            f"Expected a Runnable, callable or dict."
            f"Instead got an unsupported type: {type(thing)}"
        )


def _is_optional_type(type_: Any) -> bool:
    """Check if a type is Optional."""

    if hasattr(type_, "__origin__") and hasattr(type_, "__args__"):
        origin = get_origin(type_)
        if origin is Optional:
            return True
        if origin is Union:
            return any(
                arg is type(None) or _is_optional_type(arg) for arg in type_.__args__
            )
        return origin is None
    if hasattr(type_, "__bound__") and type_.__bound__ is not None:
        return _is_optional_type(type_.__bound__)
    return type_ is None


def _is_required_type(type_: Any) -> Optional[bool]:
    """Check if an annotation is marked as Required/NotRequired.

    Returns:
        - True if required
        - False if not required
        - None if not annotated with either
    """
    origin = get_origin(type_)
    if origin is Annotated or origin:
        # See https://typing.readthedocs.io/en/latest/spec/typeddict.html#interaction-with-annotated
        return _is_required_type(type_.__args__[0])
    if origin is Required:
        return True
    if origin is NotRequired:
        return False
    return None


def _is_readonly_type(type_: Any) -> bool:
    """Check if an annotation is marked as ReadOnly.

    Returns:
        - True if is read only
        - False if not read only
    """

    # See: https://typing.readthedocs.io/en/latest/spec/typeddict.html#typing-readonly-type-qualifier
    origin = get_origin(type_)
    if origin is Annotated:
        return _is_readonly_type(type_.__args__[0])
    if origin is ReadOnly:
        return True
    return False


_DEFAULT_KEYS = frozenset()


def field_is_optional(name: str, type_: Any, schema: Type[Any]) -> bool:
    """Determine if the field is optional for a graph's input.

    This is based on:
        If TypedDict:
            - Required/NotRequired
            - total=False -> everything optional
        - Type annotation (Optional/Union[None])
    """
    optional_keys = getattr(schema, "__optional_keys__", _DEFAULT_KEYS)
    if name in optional_keys:
        # Either total=False or explicit NotRequired.
        # No type annotation trumps this.
        return True
    if _is_required_type(type_):
        # Handle Required[<type>]
        # (we already handled NotRequired and total=False)
        return False
    # Note, we ignore ReadOnly attributes,
    # as they don't make much sense. (we don't care if you mutate the state in your node)
    # and mutating state in your node has no effect on our graph state.
    # Base case is the annotation
    return _is_optional_type(type_)
