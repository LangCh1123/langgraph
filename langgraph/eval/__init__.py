from collections import defaultdict
from typing import Any, Dict, Iterable, List, TypedDict

from langchain_core.runnables import RunnableConfig

from langgraph.pregel import Pregel
from langgraph.pregel.types import StateSnapshot


class TestCase(TypedDict):
    id: str
    inputs: Dict[str, Any]
    outputs: Dict[str, Any]
    metadata: Dict[str, Any]


def _node_test_cases(snapshots: Iterable[StateSnapshot]) -> Dict[str, List[TestCase]]:
    test_cases = defaultdict(list)
    partials: Dict[str, Dict[str, TestCase]] = defaultdict(dict)
    for snapshot in snapshots:
        thread_ts = snapshot.config["configurable"]["thread_ts"]
        if partials[thread_ts]:
            for node, partial in partials[thread_ts].items():
                test_cases[node].append(
                    {
                        "id": partial["id"],
                        "inputs": snapshot.values,
                        "outputs": partial["outputs"],
                        "metadata": partial["metadata"],
                    }
                )
            partials[thread_ts].clear()
        if (
            (writes := snapshot.metadata["writes"])
            and snapshot.parent_config
            and isinstance(writes, dict)
            and snapshot.metadata["source"] == "loop"
        ):
            parent_thread_ts = snapshot.parent_config["configurable"]["thread_ts"]
            for node, outputs in writes.items():
                partials[parent_thread_ts][node] = {
                    "id": snapshot.config["configurable"]["thread_ts"],
                    "inputs": None,
                    "outputs": outputs,
                    "metadata": {
                        "source": snapshot.metadata["source"],
                        "step": snapshot.metadata["step"],
                        **snapshot.config["configurable"],
                    },
                }
    return dict(test_cases)


def extract_node_test_cases_from_thread(
    graph: Pregel, config: RunnableConfig
) -> Dict[str, List[TestCase]]:
    return _node_test_cases(graph.get_state_history(config))


async def aextract_node_test_cases_from_thread(
    graph: Pregel, config: RunnableConfig
) -> Dict[str, List[TestCase]]:
    return _node_test_cases([s async for s in graph.get_state_history(config)])


def _graph_test_case(snapshots: Iterable[StateSnapshot]) -> TestCase:
    test_case = TestCase(
        id=None,
        inputs={
            "input": [],
        },
        outputs={
            "output": [],
            "steps": [],
        },
    )
    is_acc_steps = False
    for snapshot in snapshots:
        if not test_case["id"]:
            test_case["id"] = snapshot.config["configurable"]["thread_id"]
        if not snapshot.next:
            is_acc_steps = True
            test_case["outputs"]["output"].append(snapshot.values)
            test_case["outputs"]["steps"].append([])
            if not test_case.get("metadata"):
                test_case["metadata"] = snapshot.config["configurable"]
        if (
            is_acc_steps
            and snapshot.metadata["source"] == "loop"
            and snapshot.metadata["writes"]
        ):
            for node in snapshot.metadata["writes"]:
                test_case["outputs"]["steps"][-1].append(node)
        if is_acc_steps and snapshot.metadata["source"] == "input":
            test_case["inputs"]["input"].append(snapshot.metadata["writes"])
    test_case["inputs"]["input"].reverse()
    test_case["outputs"]["output"].reverse()
    test_case["outputs"]["steps"].reverse()
    for ss in test_case["outputs"]["steps"]:
        ss.reverse()
    return test_case


def extract_graph_test_case_from_thread(
    graph: Pregel, config: RunnableConfig
) -> TestCase:
    return _graph_test_case(graph.get_state_history(config))


async def aextract_graph_test_case_from_thread(
    graph: Pregel, config: RunnableConfig
) -> TestCase:
    return _graph_test_case([s async for s in graph.get_state_history(config)])
