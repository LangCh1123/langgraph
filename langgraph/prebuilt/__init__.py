"""langgraph.prebuilt exposes a higher-level API for creating and executing agents and tools."""

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from langgraph.prebuilt import chat_agent_executor
    from langgraph.prebuilt.agent_executor import create_agent_executor
    from langgraph.prebuilt.chat_agent_executor import create_react_agent
    from langgraph.prebuilt.tool_executor import ToolExecutor, ToolInvocation
    from langgraph.prebuilt.tool_node import ToolNode, tools_condition
    from langgraph.prebuilt.tool_validator import ValidationNode


def __getattr__(name: str) -> Any:
    if name == "chat_agent_executor":
        from langgraph.prebuilt import chat_agent_executor

        return chat_agent_executor
    if name == "create_agent_executor":
        from langgraph.prebuilt.agent_executor import create_agent_executor

        return create_agent_executor
    if name == "create_react_agent":
        from langgraph.prebuilt.chat_agent_executor import create_react_agent

        return create_react_agent
    if name == "ToolExecutor":
        from langgraph.prebuilt.tool_executor import ToolExecutor

        return ToolExecutor
    if name == "ToolInvocation":
        from langgraph.prebuilt.tool_executor import ToolInvocation

        return ToolInvocation
    if name == "ToolNode":
        from langgraph.prebuilt.tool_node import ToolNode

        return ToolNode
    if name == "tools_condition":
        from langgraph.prebuilt.tool_node import tools_condition

        return tools_condition
    if name == "ValidationNode":
        from langgraph.prebuilt.tool_validator import ValidationNode

        return ValidationNode
    raise AttributeError(f"module {__name__} has no attribute {name}")


__all__ = [
    "create_agent_executor",
    "chat_agent_executor",
    "create_react_agent",
    "ToolExecutor",
    "ToolInvocation",
    "ToolNode",
    "tools_condition",
    "ValidationNode",
]
