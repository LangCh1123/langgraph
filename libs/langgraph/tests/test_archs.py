def test_quickstart(snapshot) -> None:
    from typing import Annotated, TypedDict

    from langgraph.graph import END, START, StateGraph
    from langgraph.graph.message import add_messages

    class State(TypedDict):
        # Messages have the type "list". The `add_messages` function
        # in the annotation defines how this state key should be updated
        # (in this case, it appends messages to the list, rather than overwriting them)
        messages: Annotated[list, add_messages]

    builder = StateGraph(State)

    from langchain_openai import ChatOpenAI

    model = ChatOpenAI()

    def chatbot(state: State):
        answer = model.invoke(state["messages"])
        return {"messages": [answer]}

    # The first argument is the unique node name
    # The second argument is the function or Runnable to run
    builder.add_node("chatbot", chatbot)

    builder.add_edge(START, "chatbot")
    builder.add_edge("chatbot", END)

    graph = builder.compile()

    assert graph.get_graph().draw_mermaid() == snapshot
