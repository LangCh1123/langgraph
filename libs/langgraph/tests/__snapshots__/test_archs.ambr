# serializer version: 1
# name: test_quickstart
  '''
  %%{init: {'flowchart': {'curve': 'linear'}}}%%
  graph TD;
  	__start__([__start__]):::first
  	chatbot(chatbot)
  	__end__([__end__]):::last
  	__start__ --> chatbot;
  	chatbot --> __end__;
  	classDef default fill:#f2f0ff,line-height:1.2
  	classDef first fill-opacity:0
  	classDef last fill:#bfb6fc
  
  '''
# ---
